### PR TITLE
HAProxy protocol header support

### LINKED ISSUE #8757: Support HAProxy protocol header on HTTPS listener
When LXD is being run behind a HAProxy server (common with clusters), it'd be quite useful for LXD to still capture the requestor's IP address. For that, the easiest is to add an option to specify upstream haproxy servers and if a connection comes from those, then expect the proxy header and replace the client's address with it.

As this is for `core.https_address`, we'd probably want `core.https_trusted_proxy` which takes a comma separated list of IP addresses of servers that we trust to provide the client's address through the proxy connection header.

Then in the REST API endpoint, if the connection is coming from one of those addresses, we'll read the first line of data, parse the proxy header and replace the address in the request with the extracted one (there's likely a Go package doing that already, I remember implementing this in https://github.com/nsec/askgod a while back).

### PR DESCRIPTION
**Pull request by presztak**
Resolves #8757

@stgraber, this is implemented a little bit different then your implementation in https://github.com/nsec/askgod. I used `proxyproto.NewConn()` method because we need to know incoming remote address to check is it in `core.https_trusted_proxy` config. I'm also not quite sure if I'm not miss something. I mean, I update proxy address when server is starting and also when network or cluster listeners are changed. `proxyproto.newConn()` as a second argument takes timeout value (ProxyHeaderTimeout). This value should by configurable or should we use some static value?

### CODE REVIEW

**Comment by tomponline** [comment]
jenkins: test this please

File: lxd/endpoints/network.go [hunk 1]
```diff
@@ -4,9 +4,11 @@ import (
 	"crypto/tls"
 	"fmt"
 	"net"
+	"strings"
 	"sync"
 	"time"
 
+	"github.com/armon/go-proxyproto"
```
Comment (stgraber) [inline lxd/endpoints/network.go:11]: line break after this one as it's an external import so should be separate from our internal ones.

File: lxd/endpoints/network.go [hunk 2]
```diff
@@ -164,8 +185,9 @@ func networkCreateListener(address string, cert *shared.CertInfo) (net.Listener,
 // continue using the old configuration.
 type networkListener struct {
 	net.Listener
-	mu     sync.RWMutex
-	config *tls.Config
+	mu           sync.RWMutex
+	config       *tls.Config
+	trustedProxy string
```
Comment (stgraber) [inline lxd/endpoints/network.go:190]: I think it'd be more efficient if `trustedProxy` was a `[]net.IP`.

So you'd do the parsing of the string once in `NetworkUpdateTrustedProxy` and then record the list of proxy servers as Go IP addresses making `isProxy` much cheaper.

**Comment by stgraber** [comment]
@presztak approach looks good to me, this should work fine!
Just one minor nit (import statements) and I think we can make things perform a bit better by doing the IP parsing directly in `NetworkUpdateTrustedProxy`.

**Comment by stgraber** [comment]
Oops, forgot to ask for an API extension in there, will add one and then test+merge this.

**Comment by stgraber** [comment]
jenkins: test this please

**Pull request merged**